ENH: improve runtime detection of CPU features

seiko2plus opened the pull request:
### This pull-request aims to improve the runtime detection of CPU features.

> #### The cause:
>
> The current CPU detection code has limited supports for X86 features  and it's count on compiler built-in functions that not widely supported by other compilers or platforms.

#### The solution:

Implementing an independent API similar to GCC built-in functions, so instead of `__builtin_cpu_init`, `__builtin_cpu_supports` its provide `npy_cpu_init`, `npy_cpu_have` and `NPY_CPU_HAVE`.

##### For X86:

Detect almost all `X86` features via instruction `CPUID` also check OS support for `AVX`/`AVX512` , and provides  CPU feature groups that gather several features. e.g. `AVX512_KNM` detect Knights Mill's `AVX512` features.

##### For IBM/Power:

Only supports `linux` and count here on `glibc(getauxval)` to detect `VSX` support and fail-back to the compiler definitions for other platforms. 

##### For ARM:

Same as IBM/Power but its parse `/proc/self/auxv` if `glibc(getauxval)` isn't available.

#### NOTES:

- `npy_cpu_supports` is removed rather than deprecated, use the macro `NPY_CPU_HAVE(FEATURE_NAME_WITHOUT_QUOTES)` instead.

- New attribute` __cpu_features__ `added to umath module, its a dictionary contains all supported CPU feature names with runtime availability.

charris commented on numpy/core/src/common/npy_cpu.c:
```diff
@@ -0,0 +1,219 @@
+void npy_cpu_init(void)
+{
+    memset(cpu_features, 0, sizeof(cpu_features[0]) * NPY_OPT_MAX);
+    fprintf(stderr, "NumPy: Unsupported architecture\n");
```
If this is runtime, raise a python error, results needs to be checked in call. Why not just set the python error and return -1 on error, 0 otherwise and handle error on return?

charris replied:
Also above. No C print statements except for debug code.

seiko2plus replied:
I removed it instead, set all features to zero is more than enough but we will need to add python level functions to check availability of CPU features and also to disable/enable specific feature on runtime.

charris commented on numpy/core/src/common/npy_cpu.c:
```diff
@@ -0,0 +1,219 @@
+    memset(cpu_features, 0, sizeof(cpu_features[0]) * NPY_OPT_MAX);
+    fprintf(stderr, "NumPy: Unsupported architecture\n");
+}
+#endif
```
Needs a line return to end file.

charris commented:
@rgommers Just a heads up.

charris commented:
@seiko2plus I assume that you will be reusing some code from cupy, which is OK because the cupy MIT license is compatible with the BSD license we use. Just want to make sure you are aware of potential license issues down the line in case it comes up.

seiko2plus commented:
@charris,
> I assume that you will be reusing some code from cupy, which is OK because the cupy MIT license is compatible with the BSD license we use.

There no intention to reuse any code from Cupy or at least for now maybe I should take a look there , however this patch inspired by OpenCV.

> Just want to make sure you are aware of potential license issues down the line in case it comes up.

sure, I do.

seiko2plus commented on numpy/core/src/multiarray/multiarraymodule.c:
```diff
@@ -4744,6 +4744,8 @@ PyMODINIT_FUNC init_multiarray_umath(void) {
         goto err;
     }
+    /* Initialize CPU features */
+    npy_cpu_init();
```
`npy_cpu_have` requires to call `npy_cpu_init` first, Should we add it on top? also is there other uncovered areas?

seiko2plus replied:
`npy_cpu_init` must be called before `InitOperators`

juliantaylor replied:
hm one could just `cpu_init` set a static variable and have `cpu_have` call it when needed

rgommers commented on numpy/core/include/numpy/npy_cpu.h:
```diff
@@ -115,4 +116,83 @@
+{
+    NPY_OPT_NONE = 0,
+    /* X86 */
+    NPY_OPT_MMX               = 1,
```
Is this a standard approach you took from somewhere else? The numbers do feel a bit magical......

seiko2plus replied:
I preferred to follow OpenCV style for obvious reasons, feature's values should be unique and unchangeable because we gonna use it later as entries to check/disable/enable dispatch CPU features on runtime via environment variables or via python API, also the gap between values could be useful for verify entries, i.e `(NPY_OPT_NONE > x && x < PY_OPT_VSX) == X86`

rgommers replied:
Okay, so this is what OpenCV does and it's been proven. Perhaps it just needs a bit more commenting above with context and how these are used. The bit shifting code for these values isn't exactly the easiest to understand when reading for the first time.

seiko2plus replied:
I think you're right but even though the reader have to read definition of `npy_cpu_have` and `npy_cpu_init` to fully understand how theses are used, BTW usually there's no need to use it directly, it always gonna used behind the seen inside macros or functions, i.e in this patch all you need is just call macro `NPY_CPU_HAVE(FEATURE NAME)`, NPY_CPU_HAVE(AVX2) instead of `npy_cpu_have(NPY_OPT_AVX2)`.

however I will try to improve documentation on a separate patch  but after I get done from implement universal intrinsics.

by the way I'm made a mistake in here I didn't mean to leave a gap between `NPY_OPT_SSE41` and `NPY_OPT_POPCNT`, I'm going to fix it

seiko2plus replied:
For the record, the gap between feature IDs left only for ABI compatibility and there's no intention to use it from outside. Only strings.

juliantaylor commented on numpy/core/include/numpy/npy_cpu.h:
```diff
@@ -115,4 +116,83 @@
+
+enum
+{
+    NPY_OPT_NONE = 0,
```
do we want to make cpu detection part of our api?
could be useful for other projects

though the name `NPY_OPT` is not really clear enough, maybe `NPY_CPU_FEATURE`

seiko2plus replied:
> do we want to make cpu detection part of our api?
 could be useful for other projects

yes, not even cpu detection part but also `universal intrinsics`

> though the name NPY_OPT is not really clear enough, maybe NPY_CPU_FEATURE

I agree I like `NPY_CPU_{FEATURE}` more but the prefix `NPY_CPU_` already taken by CPU names, so I'm going to use `NPY_CPU_FEATURE_` as a prefix instead.

mattip commented on numpy/core/include/numpy/npy_cpu.h:
```diff
@@ -115,4 +116,84 @@
+extern "C" {
+#endif
+
+enum
```
style nit: we don't mention enums in our style guide. We seem to have both the forms
```
enum NPY_MY_ENUM {
    ...
};
```

and 
```
typedef enum  {
    ...
} NPY_MY_ENUM;
```
It would be good to add a name to this enum using one of those.

mattip commented on numpy/core/include/numpy/npy_cpu.h:
```diff
@@ -115,4 +116,84 @@
+
+#ifdef __cplusplus
+}
+#endif
```
This new code is only used inside `numpy/core/common`, right? If not required, it would be better to move this to a header there

mattip replied:
On the other hand, it would be nice to allow use from other libraries like scipy

seiko2plus replied:
> This new code is only used inside numpy/core/common, right? If not required, it would be better to move this to a header there

no, it has actually been used under the hood in every where by gh-13516. e.g. umath genreator and _simd module.

> On the other hand, it would be nice to allow use from other libraries like scipy

Definitely yes.

mattip replied:
Lesson learned from `numpy.random` refactor: I think it would be prudent to start with this not in the public headers for the first release. The `_simd` module can reach into numpy internals to get the headers, and `umath_generator` generates code, it does not compile it, so I don't see how they require public headers.

Once this is merged and we have a clear picture of how all the parts work together, we can figure out what needs to be publicly exposed.

seiko2plus replied:
It makes sense to me. thanks!

rgommers commented:
This PR conflicts with gh-13516 pretty badly, I can't combine the two. This PR seems useful in its own right, independent of what we end up doing in gh-13516. @seiko2plus could you comment on the interaction between the two? Why doesn't gh-13516 rely on this PR for example? Should we merge this first and then rebase gh-13516 on it?

mattip commented on numpy/core/include/numpy/npy_cpu.h:
```diff
@@ -39,6 +39,7 @@
      * _M_AMD64 defined by MS compiler
      */
     #define NPY_CPU_AMD64
+    #define NPY_CPU_X86_64
```
Why do we need both defines?

seiko2plus replied:
Just an alias to make everyone happy, I used the term "X86_64" in all infrastructure work for x86 64bit mode but it's okay I can remove it and use only `NPY_CPU_AMD64` since the current definition `NPY_CPU_AMD64` already been used in another header `numpy/core/include/numpy/npy_endian.h`  .

mattip commented:
This should have some tests before we merge it. We have CI runs on aarch64 and s390x, as well as x86, so we should be able to check most of the detection code (depending on features available in th CPUs we run CI o)n.

seiko2plus commented:
@rgommers, Alright I updated this pr with the latest code from gh-13516, also I separated it into small pieces to make it easy to review. gh-13516 required this pr to be merged first and it already has these patches as a temporary commit, going to rebase it later after this pr get merged.

seiko2plus commented:
@mattip, yes, it should have some tests but I can't achieve it within this pr but I can cover it in gh-13516. I think the only practical way is to test NumPy attribute `__cpu_features__`  against /proc/cpuinfo which limits testing support for Linux only.
NumPy attribute `__cpu_features__`  is a new attribute implmented by gh-13516, contains all CPU features with their runtime check and it counts on this pr to detect the  CPU features.

mattip commented on numpy/core/src/common/npy_cpu.c:
```diff
@@ -0,0 +1,370 @@
+    if (feature_id <= NPY_CPU_FEATURE_NONE || feature_id >= NPY_CPU_FEATURE_MAX)
+        return 0;
+    return npy__cpu_have[feature_id];
+}
```
This function can be tested in the PR:
- that the set of features only appear on this CPU (easy)
- that `/proc/cpu_info` matches the set of valid features (harder)

You can write C level functions in `numpy/core/src/umath/_umath.tests.c.src and then test them from python.

seiko2plus replied:
> that the set of features only appear on this CPU (easy)

Testing against `GCC` built-in functions will limit the support for X86 also `__builtin_cpu_supports` don't check OS support.

> that /proc/cpu_info matches the set of valid features (harder)

The only way I guess but it may fail on ARM if the Linux kernel is badly patched, in ` PPC64` we rely on auxiliary vector via the `LD_SHOW_AUXV` environment variable since `/proc/cpuinfo` doesn't provide information about `VSX`.

mattip commented on numpy/core/src/multiarray/multiarraymodule.c:
```diff
@@ -4567,6 +4567,11 @@ PyMODINIT_FUNC init_multiarray_umath(void) {
+    if (npy_cpu_init() < 0) {
+        return RETVAL(NULL);
+    }
+
```
This should be closer to the `s = npy_cpu_features_dict();` call below

seiko2plus replied:
`npy_cpu_init()` must be called before `InitOperators()` since `npy_cpu_have` is used by umath generator

mattip replied:
ok, then move it closer to that call. It does not need to be before the module creation code.

seiko2plus replied:
currently, it doesn't return any errors but within the second pull request it will return the following errors:
- the validation errors of disabling CPU features via the environment variable.
- the validation errors of the baseline feature availability.

so it makes more sense to call it before ushering the module.

mattip replied:
leaving this for now as is

mattip reviewed (commented):
+1 for the test. This looks good, just a few style things and I think it is ready to go in as a replacement for the current cpu feature checking. I am not convinced we need to expose `__cpu_features__` as part of `__all__`. Usually we name private things with a single underscore: `_cpu_features`

seiko2plus commented on numpy/core/setup_common.py:
```diff
@@ -135,11 +135,6 @@ def check_api_version(apiversion, codegen_dir):
-                       ("volatile int r = __builtin_cpu_supports", '"sse"',
-                        "stdio.h", "__BUILTIN_CPU_SUPPORTS"),
-                       ("volatile int r = __builtin_cpu_supports", '"avx512f"',
-                        "stdio.h", "__BUILTIN_CPU_SUPPORTS_AVX512F"),
```
`__builtin_cpu_supports` isn't involved in this patch and it seems a workaround that apply to the older versions of clang. I don't have access to `OSX 10.11` right now.

seiko2plus commented on numpy/core/src/common/npy_cpu_features.c.src:
```diff
@@ -0,0 +1,385 @@
+    __asm__(NPY__CPUID_ASM : "=a" (reg[0]), "=b" (reg[1]), "=c" (reg[2]), "=d" (reg[3]) : "a" (func_id), "c" (0) : );
+#else
+    // TODO: handle other x86 compilers
+    reg[0] = 0;
```
honestly, I'm not sure but maybe c/c++builder(broland). IBM/XLC doesn't support x86 and it based on clang.

seiko2plus commented on numpy/core/src/umath/cpuid.c:
```diff
@@ -1,97 +0,0 @@
-#endif
-    }
-    else if (strcmp(feature, "fma") == 0) {
-        return cpu_supports_fma() && __builtin_cpu_supports("avx2") && os_avx_support();
```
Why `FMA` require `AVX2` instead of `AVX`? any reason?

seiko2plus replied:
Well I got it, the current  SIMD `FMA3` code requires `AVX2` but now I don't understand the point behind checking the availability of `FMA3` and `AVX2`. Is there any CPU that supports `AVX2` and at the same time doesn't support `FMA3`?!!!.

seiko2plus replied:
> Where do we have any backward compatibility problems?

/cc @mattip, here

seiko2plus replied:
For safety `FMA` is no longer alias for `AVX2` but its an independent feature detecting `AVX2` and `FMA3` together.

r-devulap replied:
> Is there any CPU that supports `AVX2` and at the same time doesn't support `FMA3`?!!!.

Example of a CPU with FMA3 and no AVX2: https://en.wikipedia.org/wiki/AMD_FX 
I know that's a discontinued product, but in general, if a feature has its own CPUID flag, I would think its safer to use them and not make assumptions.

seiko2plus replied:
@r-devulap, I was meaning that all CPUs that support `AVX2` also supports `FMA3` but not the opposite, so checking only `AVX2` will be enough and that's what almost all projects do but after a deep search, I found new X86 chips re-designed by `VIA` support `AVX2` but without `FMA3` supports.
so currently I'm working on improving the infrastructure in gh-13516, it almost re-designed in order to be extremely flexible so it can fix this issue as follows:

- `AVX2` is no longer implies `FMA3` and both are separated  but in Intel compilers
  this behavior will change to force `FMA3` and `AVX2` to imply each other since there are no
  sperate flags for them only `-march=core-avx2` and `/arch:core-avx2`, same for features `AVX512F` and `AVX512CD`

- the new infrastructure configurable syntax must support multiple dispatch-able features in the same object, Example:
```` 
/*@targets FMA3 AVX2 (FMA3 AVX2) */ // parentheses() used for multiple targets
````
the above config will generate the follows:
- an object for `FMA3`, with requesting a runtime detect for `FMA3`
- an object for `AVX2`, with requesting a runtime detect for `AVX2`
- an object for `AVX2` and `FMA3`,  with requesting a runtime detect for `AVX2` and `FMA3`
the infrastructure will be dynamic enough to deal with the compiler limitations scenarios to avoid creating duplicated objects in case of intel compiler

I will update the gh-13516 soon with the new updates then I open it for review instead of the current draft status, also I want to add that I put all your previous comments into high considerations but some of them need to be replied within a code with a benchmark that proves the efficiency behind all this work. hopefully, at the end, I can make everyone get satisfied.

seiko2plus commented on numpy/core/src/common/npy_cpu_features.c.src:
```diff
@@ -0,0 +1,397 @@
+    PyObject *dict = PyDict_New();
+    if (dict) {
+    /**begin repeat
+     * #feature = MMX, SSE2, SSE3, SSSE3, SSE41, POPCNT, SSE42, AVX, F16C, XOP, FMA4, FMA3, AVX2, AVX512F,
```
typo, `SSE` is missing.

seiko2plus commented on numpy/core/tests/test_cpu_features.py:
```diff
@@ -0,0 +1,112 @@
+
+class X86Features(AbstractTest):
+    features = [
+        "MMX", "SSE2", "SSE3", "SSSE3", "SSE41", "POPCNT", "SSE42", "AVX", "F16C",
```
`SSE` is missing here too

mattip commented on numpy/core/tests/test_cpu_features.py:
```diff
@@ -0,0 +1,112 @@
+        the_class = "class Test_{arch}_features({class_name}):\n\tpass".format(
+            class_name=class_name, arch=arch
+        )
+        exec(the_class)
```
Refactor to use a `@pytest.mark.skipif(not arch=='X86') decorator rather than this custom `exec` code

mattip commented on numpy/core/src/common/npy_cpu_features.h:
```diff
@@ -0,0 +1,116 @@
+    NPY_CPU_FEATURE_XOP               = 12,
+    NPY_CPU_FEATURE_FMA4              = 13,
+    NPY_CPU_FEATURE_AVX2              = 14,
+    NPY_CPU_FEATURE_FMA               = 14, // alias to provide backward compatibility
```
Where do we have any backward compatibility problems?

seiko2plus replied:
Here the situation,
AMD chips "Piledriver" has support for `FMA3` but without `AVX2` support.
All intel and AMD chips that support `AVX2`  also support `FMA3`.

And the current SIMD code requires `AVX2` and `FMA3` with a feature name `FMA`, so I had to add `FMA` for combability, and it seems I was wrong since mapping `FMA` to `AVX2` will not be safe because according to "Peter Cordes" in his comment at [stackoverflow.](https://stackoverflow.com/a/55763975), "there is a real x86 CPU that supports AVX2 but without FMA3"  manufactured by "VIA".

So now `FMA` is added for compatibility reasons and it detects both `AVX2` and `FMA3`.

mattip commented on numpy/core/src/common/npy_cpu_features.h:
```diff
@@ -0,0 +1,116 @@
+    // Cannon Lake     (F,CD,BW,DQ,VL,IFMA,VBMI)
+    NPY_CPU_FEATURE_AVX512_CNL        = 105,
+    // Ice Lake        (F,CD,BW,DQ,VL,IFMA,VBMI,VNNI,VBMI2,BITALG,VPOPCNTDQ)
+    NPY_CPU_FEATURE_AVX512_ICL        = 106,
```
These are not "features" but "groups", perhaps the name should reflect that?

seiko2plus replied:
group names already have extra '_'

seiko2plus replied:
@mattip, would it be better to handle CPU detections of feature groups from infrastructure instead?
During my current work in cleaning up the `ccompiler_opt.py(CCompilerOpt)` in gh-13516,
I found it will be better if we just handle detections of feature groups from `CCompilerOpt` directly,
it can be done by generating multiple requests for detecting certain features via the abstract macros that defined by the generated C headers during parse the dispatch-able sources.
That will give great flexibility in the future. what do you think?

mattip replied:
Could you create a different PR with the design you have in mind? I am not sure what you mean.

seiko2plus replied:
@mattip, working on it

mattip commented:
Perhaps we should put this in, independently of NEP 37. It is a replacement for the current cpu feature detection code, and does not require us to commit to universal intrinsics. The general concept is solid, there  some cleanup that could be done in follow-on PRs. @rgommers any thoughts?

seiko2plus committed:
ENH: improve runtime detection of CPU features

  - Put the old CPU detection code to rest

    The current CPU detection code only supports x86 and
    it's count on compiler built-in functions that not widely supported
    by other compilers or platforms.

    NOTE: `npy_cpu_supports` is removed rather than deprecated,
     use the macro `NPY_CPU_HAVE(FEATURE_NAME_WITHOUT_QUOTES)` instead.

  - Initialize the new CPU features runtime detector

    Almost similar to GCC built-in functions,
    so instead of `__builtin_cpu_init`, `__builtin_cpu_supports`
    its provide `npy_cpu_init`, `npy_cpu_have` and `NPY_CPU_HAVE`.

    NOTE: `npy_cpu_init` must be called before any use of
    `npy_cpu_have` and `NPY_CPU_HAVE`, however `npy_cpu_init`
    already called during the load of module `umath`
    so there's no reason to call it again in most of the cases.

  - Add X86 support

    detect almost all x86 features, also provide
    CPU feature groups that gather several features.
    e.g. `AVX512_KNM` detect Knights Mill's `AVX512` features

  - Add IBM/Power support

    only supports Linux and count here on `glibc(getauxval)`
    to detect VSX support and fail-back to the compiler definitions
    for other platforms.

  - Add ARM support

    Same as IBM/Power but its parse `/proc/self/auxv`
    if `glibc(getauxval)` isn't available.

  - Update umath generator

  - Add testing unit for Linux only

  - Add new attribute `__cpu_features__` to umath module

    `__cpu_features__` is a dictionary contains all supported
    CPU feature names with runtime availability

seiko2plus commented:
> Perhaps we should put this in, independently of NEP 37. It is a replacement for the current cpu feature detection code, and does not require us to commit to universal intrinsics. The general concept is solid, there some cleanup that could be done in follow-on PRs. 

@mattip, IMHO two NEPS will be better, one for infrastructure (runtime & build) and another one independent for universal intrinsics

mattip commented:
Can we merge this right now? @seiko2plus's [comment](https://github.com/numpy/numpy/pull/13421#discussion_r368976263) above seems to hint that there is a better approach. Is the intention to rewrite this entirely differently or incrementally improve it?

seiko2plus commented:
@mattip,  yes sure but just give me a minute let me fix typo `boolen` first,
> comment above seems to hint that there is a better approach

most of the changes will be inside in gh-13516, and there will be no rewriting for the detection code, yes I can tell incrementally improvements.

mattip committed:
DOC: add a release note

mattip commented:
added a release note, and fix documentation building. I intend to merge this in a day or two if there are no additional review comments.

rgommers reviewed (approved):
+1 for merging this, it feels like a nice improvement that's valuable independent of @seiko2plus's other PRs.

rgommers commented:
This is also a good time for merging, with the next release still being a couple of months off.

eric-wieser commented on numpy/core/src/common/npy_cpu_features.c.src:
```diff
@@ -0,0 +1,401 @@
+     *            NEON, NEON_FP16, NEON_VFPV4, ASIMD, FPHP, ASIMDHP, ASIMDDP, ASIMDFHM#
+    */
+        PyDict_SetItemString(dict, "@feature@",
+            npy__cpu_have[NPY_CPU_FEATURE_@feature@] ? Py_True : Py_False);
```
Should be
```
if (PyDict_SetItemString(...) < 0) {
    return NULL;
}
```

seiko2plus replied:
That's too much testing that may slow down the module load, the only case here that makes `PyDict_SetItemString()` failing is memory allocation. so would it be better to ignore the returned value?

eric-wieser replied:
I doubt that the cost of the branch is significant compared to the cost of `malloc`ing the `"@feature@"` string. Let's not trade-off unprofiled speed for correctness.

>  failing is memory allocation. so would it be better to ignore the returned value?

If memory allocations fail, we should propagate the `MemoryError` to the caller so they can recover, not keep going and segfault.

seiko2plus replied:
>  Let's not trade-off unprofiled speed for correctness.

100% right, especially after I checked how `cpython` implement `PyDict_SetItemString()` :)

eric-wieser commented on numpy/core/src/multiarray/multiarraymodule.c:
```diff
@@ -4512,6 +4517,10 @@ PyMODINIT_FUNC PyInit__multiarray_umath(void) {
     PyDict_SetItemString(d, "__version__", s);
     Py_DECREF(s);
 
+    s = npy_cpu_features_dict();
```
```suggestion
    s = npy_cpu_features_dict();
    if (s == NULL) {
        return NULL;
    }
```
I know we haven't been very good about this in the rest of the file, but we may as well get it right for new code

seiko2plus replied:
I thought its part of NumPy code style to speed up loading process

eric-wieser replied:
As far as I'm aware it's written this way because it was easiest, not because it was best.

seiko2plus replied:
`PyInit__multiarray_umath()` needs a rework, also the module should deallocated via `Py_XDECREF` with `m` initialized to `NULL`.

seiko2plus replied:
I kept the current code as-is, just modified the new.

eric-wieser commented on numpy/core/src/multiarray/multiarraymodule.c:
```diff
@@ -4512,6 +4517,10 @@ PyMODINIT_FUNC PyInit__multiarray_umath(void) {
     Py_DECREF(s);
 
+    s = npy_cpu_features_dict();
+    PyDict_SetItemString(d, "__cpu_features__", s);
```
Same theme as above
```suggestion
    if (PyDict_SetItemString(d, "__cpu_features__", s) < 0) {
        return NULL;
    }
```

eric-wieser commented on numpy/core/src/multiarray/multiarraymodule.c:
```diff
@@ -4512,6 +4517,15 @@ PyMODINIT_FUNC PyInit__multiarray_umath(void) {
 
+    s = npy_cpu_features_dict();
+    if (s == NULL)
+        goto err;
```
Braces please

eric-wieser commented on numpy/core/tests/test_cpu_features.py:
```diff
@@ -0,0 +1,102 @@
+
+    def load_flags_auxv(self):
+        import os
+        with os.popen("LD_SHOW_AUXV=1 /bin/true") as auxv:
```
Nit: can we use `subprocess.check_output('/bin/true', env=dict(LD_SHOW_AUXV=1))` or similar here?

seiko2plus replied:
modified it as u said

eric-wieser commented on numpy/lib/__init__.py:
```diff
@@ -39,9 +39,9 @@
 from .arraypad import *
 from ._version import *
-from numpy.core._multiarray_umath import tracemalloc_domain
+from numpy.core._multiarray_umath import tracemalloc_domain, __cpu_features__
```
This feels to me like it belongs in core, not lib. Would appreciate the opinion of other maintainers more involved with this.

mattip replied:
Maybe we could leave it as a private feature for now, so anyone who wants it will have to import it specifically. Then we could just delete these changes

seiko2plus replied:
alright, CPU features dictionary is no longer exposed, it became private within attr '_CPU_FEATURES'. Is that okay?

seiko2plus replied:
revert back from this modification and I keep it exposed as `__ cpu_features __` until you make a final decision. to me I'm more than welcome to apply any modification you request.

mattip replied:
please delete this line, we can decide later where is the best place to expose this publically, if at all

seiko2plus replied:
It is done, thank you

rgommers replied:
I'd suggest `numpy.__cpu_features__`, would be nice to expose I think.

eric-wieser replied:
Right, my objection was that the choice was between
* `numpy.lib.__cpu_features__` (this comment chain)
* `numpy.__cpu_features__`
* `numpy.core.__cpu_features__`
and to me only the latter two make sense.

rgommers replied:
`numpy.core` is more or less an accident of history, it really shouldn't be or look public. So I'm basically -1 on adding anything to it.

r-devulap replied:
Is `__cpu_features__ ` currently exposed anywhere? I might find this useful to enable the transcendental validation tests added in https://github.com/numpy/numpy/pull/14048

seberg replied:
It is privately exposed through: `np.core.multiarray._multiarray_umath.__cpu_features__`; Although I would not mind moving it to the main namespace, unless we expect the growth of such functions in some direction which would warrant a new submodule.

seberg replied:
I just noticed:
```
 'AVX512BITALG': False,
 'AVX512_KNL': False,
```
I assume the spelling difference (underscore vs. no underscore) is not an accident but the way these are spelled? E.g. due to being different feature sets, vs. a specific feature flag? (if that makes sense)

seiko2plus replied:
@r-devulap, just like @seberg mentioned, its already been used in `test_cpu_features.py`.
@seberg, sure, it's not an accident with underscore it means a group of CPU features,  by `AVX512_KNL` mean the AVX512 CPU features of Intel Knights Landing, the shortcuts are based on Intel compilers and SDE

r-devulap replied:
I think underscore exists for keys that represent CPU itself (SKX, KNL, CLX, CNL, ICL etc.) rather than an instruction set (AVX, AVX2, FMA, AVX512F, etc.).

mattip commented on numpy/lib/__init__.py:
```diff
@@ -39,9 +39,9 @@
+from numpy.core._multiarray_umath import tracemalloc_domain, __cpu_features__
 
-__all__ = ['emath', 'math', 'tracemalloc_domain', 'Arrayterator']
+__all__ = ['emath', 'math', 'tracemalloc_domain', 'Arrayterator', '__cpu_features__']
```
also revert this change

mattip commented:
Thanks @seiko2plus

rgommers commented:
Great to see this merged, thanks @seiko2plus!

seiko2plus commented:
wow finally thank you all :)

seiko2plus commented on numpy/core/src/common/npy_cpu_features.h:
```diff
@@ -0,0 +1,117 @@
+    NPY_CPU_FEATURE_AVX512VBMI2       = 43,
+    NPY_CPU_FEATURE_AVX512BITALG      = 44,
+
+    // X86 CPU Groups
```
@seberg, already explained here

Qiyu8 commented:
Thanks @seiko2plus .This PR is exactly what I need, as mentioned in https://github.com/scipy/scipy/issues/11482, Now I'm ready to optimize Numpy in ARM-based Platform.

seiko2plus commented:
@Qiyu8, please wait, there's a misunderstanding here, this pr doesn't provide any infrastructure for ARM, only runtime detection, you have to wait for #13516, thank you.

ryandesign commented:
Thank you for merging this, but the fix appears not to have been included in version 1.18.2 released last week.

rgommers commented:
@ryandesign it wasn't supposed to be included in 1.18.2. That was a bug fix only release, and this is a new enhancement. It will be in 1.19.0

ryandesign commented:
I was afraid you were going to say something like that. I do not consider it an enhancement; I consider it a bug fix. Without it, [we cannot build anything that uses numpy on macOS 10.12](https://github.com/numpy/numpy/issues/15342).

seberg commented:
The issue is that it is a fairly large change, so it is hard to be sure that there is no regression for anyone, and regressions are especially bad in bug-fix releases... We want anyone to be able to update to a new 1.18 release without really thinking about it.
